test/contrib/test_ssltransport.py

```diff
@@ -0,0 +1,393 @@
+                ssock.send(response)
+
+        self._start_server(socket_handler)
+
+    def test_start_closed_socket(self):
+        """ Errors generated from an unconnected socket should bubble up."""
+        sock = socket.socket(socket.AF_INET)
+        context = ssl.create_default_context()
+        sock.close()
+        with pytest.raises(OSError):
+            SSLTransport(sock, context)
+
+    def test_close_after_handshake(self):
+        """ Socket errors should be bubbled up """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            ssock.close()
+            with pytest.raises(OSError):
+                ssock.send(b"blaaargh")
+
+    def test_wrap_existing_socket(self):
+        """ Validates a single TLS layer can be established.  """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            assert ssock.version() is not None
+            ssock.send(
+                b"GET http://www.testing.com/ HTTP/1.1\r\n"
+                b"Host: www.testing.com\r\n"
+                b"User-Agent: awesome-test\r\n"
+                b"\r\n"
+            )
+            response = consume_socket(ssock)
+            assert response is not None
+
+    def test_ssl_object_attributes(self):
+        """ Ensures common ssl attributes are exposed """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            assert ssock.cipher() is not None
+            assert ssock.selected_alpn_protocol() is None
+            assert ssock.selected_npn_protocol() is None
+            assert ssock.shared_ciphers() is not None
+            assert ssock.compression() is None
+            assert ssock.getpeercert() is not None
+
+            ssock.send(
+                b"GET http://www.testing.com/ HTTP/1.1\r\n"
+                b"Host: www.testing.com\r\n"
+                b"User-Agent: awesome-test\r\n"
+                b"\r\n"
+            )
+            response = consume_socket(ssock)
+            assert response is not None
+
+    def test_socket_object_attributes(self):
+        """ Ensures common socket attributes are exposed """
+        self.start_dummy_server()
+
+        sock = socket.create_connection((self.host, self.port))
+        with SSLTransport(
+            sock, self.client_context, server_hostname="localhost"
+        ) as ssock:
+            assert ssock.fileno() is not None
+            test_timeout = 10
+            ssock.settimeout(test_timeout)
+            assert ssock.gettimeout() == test_timeout
```

Comment:
Can we also verify `sock.gettimeout()` was impacted here too

Reply:
Not sure what you mean here Seth, can you clarify?

The timeout is applied but never takes effect. I would also prefer to not add a test that validates the timeout kicks in as they can be flakey. (e.g set timeout to 2 seconds, sleep for 3, verify exception is raised)

Reply:
Wanted to add a verification that the underlying `SSLTransport.socket.gettimeout()` is equal to the value set within `SSLTransport.settimeout()`. Tbh it's not that important since it can be clearly verified.

Reply:
ah, I added it anyway :)